Şıhmehmet Yıldız

$511212133$

yildizsih@itu.edu.tr

MAT555E - Homework 3

In [13]:
import numpy as np
from scipy.io.matlab import loadmat
from scipy.optimize import linear_sum_assignment

# HW3

In this homework we are going to look at a hyperspectral image dataset called [Salinas](https://rslab.ut.ac.ir/data) dataset.

In [14]:
salinas = loadmat('data/Salinas_corrected.mat')['salinas_corrected']
salinas.shape

(512, 217, 204)

In [15]:
salinas_gt = loadmat('data/Salinas_gt.mat')['salinas_gt']
salinas_gt.shape

(512, 217)

## Q1 

1. Reshape and split the data into train (80%) and test (20%) data sets.
2. Apply all of the supervised ML algorithms we learned on the data and compare the results.

## Answer 1,

Import the necessary python packages for question 1.

In [16]:
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix, accuracy_score
from sklearn.neighbors import KNeighborsClassifier
from sklearn.cluster import KMeans, DBSCAN, AgglomerativeClustering
from scipy.stats import chisquare
import time

Upload the dataset and reshape it to the appropriate format for use.

In [17]:
X = salinas.reshape((512*217,204))
y = salinas_gt.reshape(512*217)

X_train, X_test, y_train, y_test = train_test_split(X,y,train_size=0.80)

In [18]:
np.unique(y)

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16],
      dtype=uint8)

### K-Nearest Neighbors (K-NN)

In [7]:
model = KNeighborsClassifier(n_neighbors=20,metric='minkowski')
t1 = time.time()
model.fit(X_train,y_train)
t2 = time.time()
print('Training time is ',t2-t1)
predicted = model.predict(X_test)
cm = confusion_matrix(y_test,predicted)

Training time is  2.607513189315796


In [8]:
rows, cols = linear_sum_assignment(cm,maximize=True)
cm = cm[rows,:]
cm = cm[:,cols]
cm

array([[10350,    68,   115,    51,    32,    87,    88,    65,   101,
           68,    85,    40,    49,    27,    37,    39,    36],
       [   11,   389,     4,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0],
       [    9,     0,   693,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0],
       [   77,     0,     0,   353,     0,     2,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0],
       [   39,     0,     0,     0,   260,     1,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0],
       [   23,     0,     0,     5,     2,   474,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0],
       [   10,     0,     0,     0,     0,     0,   796,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0],
       [    9,     0,     0,     0,     0,     0

In [9]:
model.score(X_test,y_test)

0.8955942576841727

In [10]:
chisquare(cm,axis=None)

Power_divergenceResult(statistic=1492769.7641870303, pvalue=0.0)

### Support Vector Machine

In [ ]:
#SVM
#Create the SVM model
from sklearn.svm import SVC
#svc_model = SVC(kernel = 'linear', random_state = 0)
svc_model = SVC(C=12.0,
             kernel='rbf',
             max_iter=1000)
#Fit the model for the data
t1 = time.time()
svc_model.fit(X_train, y_train)
t2 = time.time()
print('Training time is ',t2-t1)
#Make the prediction
predicted = svc_model.predict(X_test)


In [12]:
cm = confusion_matrix(y_test,predicted)
print(cm)
print(svc_model.score(X_train, y_train))
print(svc_model.score(X_test, y_test))
print(accuracy_score(predicted,y_test))
print(chisquare(cm,axis=None))

0.44845471012454574
0.4465145583007065
0.4465145583007065
[[1673   82  111 3952  434  442  179   77  191 2930  614   62  166  122
   104  140   41]
 [   2  401    9    0    0    0    0    0    0    0    0    0    0    0
     0    0    0]
 [  11    0  730    0    0    0    0    0    0    0    0    0    0    0
     0    0    0]
 [   0    0    0  405    0    1    0    0    0    1    0    0    0    0
     0    0    0]
 [  51    0    0    0  212    2    0    0    0    0    0    0    0    0
     0    0    0]
 [   5    0    0    7    1  548    1    0    0    1    1    0    0    0
     0    0    0]
 [   3    0    0    0    0    0  785    0    0    0    0    0    0    0
     0    0    0]
 [   9    0    0    0    0    0    0  710    0    0    0    0    0    0
     0    0    0]
 [  11    0    0    0    0    0    0    0   32    4   25    0    0    0
     0 2189    0]
 [   0    0    0    0    0    0    0    0    1 1244    0    0    0    0
     0    0    0]
 [  15    0    0   12    0    0    0    0 

Power_divergenceResult(statistic=490387.628999595, pvalue=0.0)

SVM with linear model,

In [ ]:
from sklearn.svm import LinearSVC
Lsvc_model = LinearSVC(multi_class='ovr',max_iter=3500)
Lsvc_model.fit(X_train,y_train)
predicted = Lsvc_model.predict(X_test)

In [ ]:
cm = confusion_matrix(y_test,predicted)
chisquare(cm,axis=None)
print(cm)
print(Lsvc_model.score(X_train, y_train))
print(Lsvc_model.score(X_test, y_test))
print(accuracy_score(predicted,y_test))
print(chisquare)

### Logistic Regression

In [ ]:
# Logistic Regression
from sklearn.linear_model import LogisticRegression, LogisticRegressionCV

LR_model = LogisticRegression(C=100,
                           tol=1e-1,
                           penalty='l2',
                           solver='newton-cg',
                           fit_intercept=False,
                           max_iter=8500)
t1= time.time()
LR_model.fit(X_train,y_train)
t2 = time.time()
print('Training time is ',t2-t1)

predicted = LR_model.predict(X_test)

In [ ]:
print(LR_model.score(X_train, y_train))
print(LR_model.score(X_test, y_test))
print(accuracy_score(predicted,y_test))
cm = confusion_matrix(y_test,predicted)
print(cm)
chisquare(cm,axis=None)

### Multinomial Regression,

In [ ]:
MR_model = LogisticRegression(solver='lbfgs',multi_class='multinomial')
t1 = time.time()
MR_model.fit(train_x, train_y)
t2 = time.time()
print('Training time is ',t2-t1)
MR_model.score(test_x, test_y)

predicted = MR_model.predict(X_test)

In [ ]:
print(MR_model.score(X_train, y_train))
print(MR_model.score(X_test, y_test))
print(accuracy_score(predicted,y_test))
cm = confusion_matrix(y_test,predicted)
print(cm)
chisquare(cm,axis=None)

### Linear Discriminant Analysis,

In [ ]:
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis as LDA
import matplotlib.pyplot as plt

plt.rcParams['figure.figsize'] = (9,9)

LDA_model = LDA(n_components=2)
t1 = time.time()
data = LDA_model.fit_transform(X_train, y_train)
t2 = time.time()
print('Training time is ',t2-t1)
plt.rcParams['figure.figsize'] = (9,9)
plt.scatter(data[:,0],data[:,1],c=y_train,cmap='jet',s=100,alpha=0.5)

In [9]:
predicted = LDA_model.predict(X_test)

In [10]:
print(LDA_model.score(X_train, y_train))
print(LDA_model.score(X_test, y_test))
print(accuracy_score(predicted,y_test))
cm = confusion_matrix(y_test,predicted)
print(cm)
chisquare(cm,axis=None)

0.8109537256843266
0.813194725709914
0.813194725709914
[[8585  104  133  398  199  482  130  124   64  409  286   91   93   95
    74  105   70]
 [   0  390    5    0    0    0    0    0    0    0    0    0    0    0
     0    0    0]
 [   3    1  700    0    0    0    0    1    0    0    0    0    0    0
     1    0    0]
 [  16    0    0  356    0    9    0    0    0    0    0    0    0    0
     0    0    0]
 [   5    0    0    0  264   30    0    0    0    0    0    0    0    0
     0    0    0]
 [   3    0    0    2    8  524    0    0    0    0    0    0    0    0
     0    0    0]
 [   6    0    0    0    0    2  793    0    0    0    0    0    0    0
     0    0    0]
 [   2    0    0    0    0    0    0  762    0    0    0    0    0    0
     0    0    0]
 [  31    0    0    0    0    0    0    0 1943    0   10    0    0    0
     0  314    0]
 [ 104    0    0    0    0    0    0    0    0 1112    0    0    1    0
     0    0    0]
 [  53    0    0    0    0    0    0    0    

Power_divergenceResult(statistic=1065699.945907025, pvalue=0.0)

## References

1. [Geeksforgeeks - SVM](https://www.geeksforgeeks.org/support-vector-machine-algorithm/)
2. [MAT-555E Lecture Notes](https://github.com/kaygun/2022-MAT555E)
3. [Scikit-Learn](https://scikit-learn.org/stable/)


## Q2

Now, apply all of the unsupervised algorithms we have learned so far and compare the results. In order to align the unsupervised labels and true labels use the `linear_sum_assignment` function from `scipy.optimize`. For example, if a confusion matrix looks like:

In [7]:
cm = np.array(
      [[1,10,2],
       [9,2,1],
       [3,4,10]])

we get

In [8]:
rows, cols = linear_sum_assignment(cm,maximize=True)
cm = cm[rows,:]
cm = cm[:,cols]
cm

array([[10,  1,  2],
       [ 2,  9,  1],
       [ 4,  3, 10]])

## Answer 2,

Unsupervised machine algortihms we have learned are,

* K-means

* Hierarchical Cluster

* DSCAN 

The application of these 3 methods for the data set, respectively, is given below.

Import the necessary python packages for question 2.

In [9]:
import numpy as np
from scipy.io.matlab import loadmat
from scipy.optimize import linear_sum_assignment
from sklearn.cluster import KMeans, DBSCAN, AgglomerativeClustering
from sklearn.metrics import confusion_matrix, accuracy_score
from scipy.stats import chisquare
import time

Upload the dataset and reshape it to the appropriate format for use.

In [10]:
salinas = loadmat('data/Salinas_corrected.mat')['salinas_corrected']
salinas.shape
salinas_gt = loadmat('data/Salinas_gt.mat')['salinas_gt']
salinas_gt.shape
X = salinas.reshape((512*217,204))
y = salinas_gt.reshape(512*217)

### K-means

K-means is unsupervised clustering algorithm using K number of clusters. The algorithm assigns data points to one of the clusters depending on lenght of the cluster center.


In [11]:
#K-means
k = len(np.unique(y)) # define cluster number
Kmeans_model = KMeans(n_clusters=k) #create model
t1 = time.time()
predicted = Kmeans_model.fit_predict(X) # train model
t2 = time.time()

print('Training time is ', t2-t1)

Training time is  10.78682804107666


In [12]:
cm = confusion_matrix(y,predicted) #calculate confusion matrix
rows, cols = linear_sum_assignment(cm,maximize=True)
cm = cm[rows,:]
cm = cm[:,cols]
print(cm)
chisquare(cm,axis=None) #chisquare test

[[12177   697  2792   902  4109  9543   604   652   789 13109  2732  5366
     66  1588   884   666   299]
 [    0  2004     5     0     0     0     0     0     0     0     0     0
      0     0     0     0     0]
 [    0  2136  1475     0     0     0     0    96     2     0     0     0
      0     0    12     0     5]
 [ 1420     0     0     0     2   316     0     0     0   238     0     0
      0     0     0     0     0]
 [    0     0     0     0  1384    10     0     0     0     0     0     0
      0     0     0     0     0]
 [   98     0     0     1    31  2543     0     0     0     0     0     5
      0     0     0     0     0]
 [    1     0     0     0     0     0  3747     0     0     3    91    95
      0     2    20     0     0]
 [    1     0     6     0     0     1     0  3471     6     0     0     0
      0     3     0     0    91]
 [    1     0     0     0     1     0     0     6  7670    10     9    64
      0   389    12  2971   138]
 [    0     0     0     0     0     0

Power_divergenceResult(statistic=1608165.94281034, pvalue=0.0)

### Hierarchical Clustering
Hierarchical clustering is an unsupervised learning method based on hierarchical ordering. The hierarchical clustering can be classified into the two different type of clustering:

* Agglomerative (bottom-up) hierarchical clustering
* Divisive (top-down) hierarchical clustering


In [13]:
# Hierarchical Cluster
k = len(np.unique(y))
HC_model = AgglomerativeClustering(n_clusters=k, 
                                linkage="complete", 
                                affinity='manhattan')

t1 = time.time()
predicted = HC_model.fit_predict(X)
t2 = time.time()
print('Training time is ', t2-t1)

Training time is  1663.1189179420471


In [15]:
cm = confusion_matrix(y,predicted) #calculate confusion matrix
rows, cols = linear_sum_assignment(cm,maximize=True)
cm = cm[rows,:]
cm = cm[:,cols]
print(cm)
chisquare(cm,axis=None) #chisquare test

[[30708   919  3957  1047  3933     1   629    21  1556  6887  5584    32
     10    22   367   924   378]
 [    0     0  2009     0     0     0     0     0     0     0     0     0
      0     0     0     0     0]
 [    0     0  3699     0     0     0     0     0    26     0     0     0
      0     0     0     1     0]
 [ 1927     0     0     0     2     0     0     0     0    47     0     0
      0     0     0     0     0]
 [   17     0     0    69  1308     0     0     0     0     0     0     0
      0     0     0     0     0]
 [ 2638     0     0     0    32     1     0     0     0     4     3     0
      0     0     0     0     0]
 [    6     0     0     0     0     0  3818     0     5     0   129     0
      0     0     1     0     0]
 [    1     0  2800     0     0     0     0     0   772     0     3     0
      0     0     0     3     0]
 [    1     4     4     0     2     0     0     0  8228    12   256     0
      0     0     0  2764     0]
 [ 2407     0     0     0     0     0

Power_divergenceResult(statistic=3124934.310825893, pvalue=0.0)

### DBSCAN

Density-based spatial clustering of noisy applications (DBSCAN) is a density-based data clustering algorithm used in data mining and machine learning. The algorithm groups 'densely grouped' data points into a cluster. It can identify clusters in large spatial datasets by looking at the local density of the data points. One of the most important features of DBSCAN clustering against other clustering algorithms is that it is robust against outliers. Also, DBSCAN does not require the number of clusters unlike K-means and Hierarchical clustering.

In [16]:
DBSCAN_model = DBSCAN()

t1 = time.time()
DBSCAN_model.fit(X)
t2 = time.time()
print('Training time is ', t2-t1)
predicted = DBSCAN_model.fit_predict(X)

Training time is  15.369641065597534


In [17]:
cm = confusion_matrix(y,predicted) #calculate confusion matrix
rows, cols = linear_sum_assignment(cm,maximize=True)
cm = cm[rows,:]
cm = cm[:,cols]
print(cm)
chisquare(cm,axis=None) #chisquare test

[[    0     0     0     0     0     0     0     0     0     0     0     0
      0     0     0     0     0     0]
 [    6 56963     6     0     0     0     0     0     0     0     0     0
      0     0     0     0     0     0]
 [    0  2009     0     0     0     0     0     0     0     0     0     0
      0     0     0     0     0     0]
 [    0  3726     0     0     0     0     0     0     0     0     0     0
      0     0     0     0     0     0]
 [    0  1976     0     0     0     0     0     0     0     0     0     0
      0     0     0     0     0     0]
 [    0  1394     0     0     0     0     0     0     0     0     0     0
      0     0     0     0     0     0]
 [    0  2678     0     0     0     0     0     0     0     0     0     0
      0     0     0     0     0     0]
 [    0  3959     0     0     0     0     0     0     0     0     0     0
      0     0     0     0     0     0]
 [    0  3579     0     0     0     0     0     0     0     0     0     0
      0     0     0   

Power_divergenceResult(statistic=10222045.201756911, pvalue=0.0)

## References

1. [How to Master the Popular DBSCAN Clustering Algorithm for Machine Learning](https://www.analyticsvidhya.com/blog/2020/09/how-dbscan-clustering-works/)
2. [Scikit-Learn web-site](https://scikit-learn.org/stable/modules/clustering.html)

## Q3

Now compare supervised and unsupervised algorithm results.

## Answer 3

#### Comparison of unsupervised machine learning algorithms (K-means, Hierarchical Clustering, DBSCAN)

Comparison of algorithm from application results,

* When the results are examined, it is seen that K-means makes the clustering as soon as possible. Hierarchical clustering is the slowest method. ($t_{k-means} \approx 10 sec,t_{HC} \approx 1663 sec, t_{DBSCAN} \approx 15 sec $)

* If the Chi-Square test results are examined, it is seen that the most successful clustering DBSCAN method does. It is seen that the worst clustering is done by the K-means method.

* Not specifying the number of clusters in the DBSCAN method may play a role in the success of DBSCAN. The chi-square test result can be improved by determining the number of best clusters in K-means and Hierarchical clustering methods. However, it probably gave the best results because of the DBSCAN algorithm.

* In the hierarchical clustering method, a memory error has been received on the personal computer and it has been observed that it uses too much memory. The other two methods could be used on the personal computer without getting a memory error. Hierarchical Clustering method has high memory usage and is a disadvantageous method in this respect.


Comparison of algorithm from algorithm's general procedure,

* DBSCAN can model outliers in data. Therefore, it is robust to outliers. On the other hand, K-means and Hierarchical clustering methods cannot handle outliers. 

* K-means is a distance-based lazy algorithm, so it is quite fast. On the contrary, Hierarchical clustering is based on hierarchical alignment, so it is a rather cumbersome (slow) method.

* K-means and Hierarchical clustering methods must specify how many clusters there will be as inputs. However, there is no need to specify the number of clusters in the DBSCAN method.

* Since DBSCAN is a density-based method, it can handle outliers as well as clusters of different shapes and sizes. However, K-means, Hierarchical clustering methods may not be able to handle clusterings of complex shapes and sizes.

* In the Hierarchical Clustering method, memory usage is proportional to $O(n^2)$ and time usage is proportional to $O(n^3)$. Therefore, the use of memory and time in high-dimensional data is increasing. In the other two methods, memory and time usage is quite low compared to hierarchial clustering.

* Determining the best number of clusters in the K means method can be complicated. It has a disadvantage in this respect. In the DBSCAN method, not specifying the number of clusters stands out as an advantage.

* The K-means algorithm is suitable for large datasets due to its simple clustering algorithm. The DBSCAN algorithm, on the other hand, may not be able to handle the clustering of large datasets. Hierarchical clustering is not suitable for large data sets in terms of memory and time.

* In the DBSCAN algorithm, on the other hand, the variance in density can negatively affect clustering. Large variance in Density can negatively affect proper clustering. On the other hand, since the method is sensitive to eps (The maximum distance between two samples for one to be considered as in the neighbourhood of the other) and min point (minimum data in a cluster) parameters, it is necessary to determine the appropriate eps and min point for the data. This can also be considered a disadvantage.

#### Comparison of supervised machine learning algorithms,

Comparison of algorithm from application results,

* When the analyzes were repeated in some of the methods used (KNN, SVM), different accuracy metrics were obtained. This situation shows that the methods are not deterministic methods, they are stochastic methods.

* KNN and LDA are very fast methods, and when accuracy metrics are compared, it is observed that the KNN method gives better results than the LDA method.

* SVR, Logistic and Multinomial Regression methods take longer computation time than KNN and LDA methods.

* The accuracy of the SVR method for the considered data set is lower than KNN and LDA.

## References
1. [K-means, DBSCAN, GMM, Agglomerative clustering — Mastering the popular models in a segmentation problem](https://towardsdatascience.com/k-means-dbscan-gmm-agglomerative-clustering-mastering-the-popular-models-in-a-segmentation-c891a3818e29)
2. [Scikit-Learn web-site](https://scikit-learn.org/stable/modules/clustering.html)
3. [Difference between K-Means and DBScan Clustering](https://www.geeksforgeeks.org/difference-between-k-means-and-dbscan-clustering/)
4. [How to cluster large amounts of data with minimal memory usage](https://stackoverflow.com/questions/58396826/how-to-cluster-large-amounts-of-data-with-minimal-memory-usage)
5. [Comparing supervised learning algorithms](https://www.dataschool.io/comparing-supervised-learning-algorithms/)
